Updating page title as dashboard loads

npfitz opened the pull request:
PR to update the document title as a dashboard loads in. Text will update as cards load in. Once everything is loaded, the favicon is changed to a blue checkmark. If the tab is in the background when everything is finished loading, we hold the checkmark and display a ready message until the user returns to the page.

One global change was to the Title.jsx HoC. I needed the loading message to appear before the loading timer, as per the requirements.
![chrome_vG8zJzKtOl](https://user-images.githubusercontent.com/1328979/160179991-fcd9851f-8488-4a4d-a98b-38c7a5b932cc.png)
![chrome_QVwplNwfyN](https://user-images.githubusercontent.com/1328979/160180005-8d5967c3-3f33-4fec-a345-a97cd6e589b4.png)

npfitz committed:
Updating page title as dashboard loads

kulyk commented on frontend/src/metabase/lib/favicon.js:
```diff
@@ -0,0 +1,22 @@
+
+const DEFAULT_FAVICON = "app/assets/img/favicon.ico";
+
+export const flashFavicon = _.debounce((icon, timeout) => {
+  document.querySelector('link[rel="icon"]').setAttribute("href", icon);
+  if (timeout) {
+    setTimeout(() => {
+      document
+        .querySelector('link[rel="icon"]')
+        .setAttribute("href", DEFAULT_FAVICON);
+    }, timeout);
+  }
+});
+
+export const resetFavicon = _.debounce((delay = 0) => {
+  setTimeout(() => {
+    document
+      .querySelector('link[rel="icon"]')
+      .setAttribute("href", DEFAULT_FAVICON);
+  }, delay);
+});
```
Why wrap functions with `_.debounce` without providing the debounce timeout as a second argument?

kulyk replied:
Also, I wonder if `setTimeout` should be better done in components using these functions:

1. If we want to customize a favicon for other pages, they won't necessarily need the timeout
2. It'd be easier for a particular component to save the timeout ID and then do `clearTimeout` before unmounting to avoid memory leaks and unexpected favicon changes. Right now, if you call `flashFavicon` with e.g. 5-sec timeout and go to another page, the favicon will change somewhere else.

npfitz replied:
> Why wrap functions with `_.debounce` without providing the debounce timeout as a second argument?

Haha, I honestly thought that maybe underscore had a default debounce timeout because of what I saw [here](https://github.com/metabase/metabase/blob/master/frontend/src/metabase/hoc/Title.jsx#L9). But I'll go ahead and add 300ms or something. The debounce may not even be super necessary, but as it's called in a componentDidUpdate I thought it might be a good diea.

> Also, I wonder if setTimeout should be better done in components using these functions

We could do that. I was thinking that having a central location where the favicon is managed may be easier, but that may not be the case. With regards to clearing the timeout, I could always return the timeout ID from the function for the caller to then use.

kulyk replied:
> Haha, I honestly thought that maybe underscore had a default debounce timeout because of what I saw [here](https://github.com/metabase/metabase/blob/master/frontend/src/metabase/hoc/Title.jsx#L9). But I'll go ahead and add 300ms or something. The debounce may not even be super necessary, but as it's called in a componentDidUpdate I thought it might be a good idea.

Oh right, got it. Yeah, I think it'd be better to have the param set explicitly to avoid confusion

kulyk commented on frontend/src/metabase/hoc/Title.jsx:
```diff
@@ -53,7 +54,12 @@ const title = documentTitleOrGetter => ComposedComponent =>
-          result.refresh.then(() => this._updateDocumentTitle());
+          result.refresh?.then(() => this._updateDocumentTitle());
+
+          //Getter can also return a priority index used for sorting the component stack
```
nip: we usually have a whitespace between `//` and the beginning of an actual comment

kulyk commented on frontend/src/metabase/dashboard/containers/DashboardApp.jsx:
```diff
@@ -74,7 +86,28 @@ const mapDispatchToProps = {
+      };
+    }
+    if (isLoadingDashCardsComplete && !hasBeenSeen) {
+      return "Your Dashboard is Ready";
```
Please use the `t` function from `ttag` to translate strings. This should look like:

```js
import { t } from "ttag";

return t`Your Dashboard is Ready`
```

kulyk commented on frontend/src/metabase/dashboard/containers/DashboardApp.jsx:
```diff
@@ -74,7 +86,28 @@ const mapDispatchToProps = {
+  }) => {
+    if (isLoadingDashCards) {
+      return {
+        title: `${cardsLoaded} / ${totalDashCards} loaded`,
```
Please use a combination of `msgid` and `ngettext` functions from the `ttag` package to make this translatable. See an example [here](https://github.com/metabase/metabase/blob/ebf1dec4786177dfb5fb56f425ac122bae143e43/frontend/src/metabase/components/MetadataInfo/TableInfo/ColumnCount.tsx#L18)

npfitz replied:
@kulyk I don't believe the string needs to be pluralized here, Could this also just be a `t` ?

kulyk replied:
Oh, yeah, I guess so

kulyk commented on frontend/src/metabase/dashboard/actions.js:
```diff
@@ -138,12 +138,16 @@ export const FETCH_DASHBOARD_PARAMETER_FIELD_VALUES =
 export const setSidebar = createAction(SET_SIDEBAR);
 export const closeSidebar = createAction(CLOSE_SIDEBAR);
 
+export const setDashboardSeen = createAction(SET_DASHBOARD_SEEN);
+
```
I wonder if we can improve the naming around the `dashboardSeen` thing? I'm worried it won't be immediately straightforward to get the idea of it if you don't know about dynamic favicons. E.g. when I just open a dashboard and watch cards loading, will it mean that I've seen the dashboard? Maybe something like `hasSeenLoadedDashboard` would suit better? Just an idea though

agnes-pinhanelli commented:
Hey! This is looking good and it is heading to the right direction. 

I have two minor tweaks about the copy: we don't need to use capital letters at the beginning of each word in that sentence, let’s just use “Your dashboard is ready” in this case. I also think we can delete the space between the numbers and the slash while it’s loading, so instead of “1 / 3 loaded” it would be “1/3 loaded”.

But this is still in progress, right? Asking because I wanted to check if we would be able to apply[ the animated indicator instead of just the favicon that is proposed in the product doc](https://www.notion.so/metabase/Reduce-the-waiting-cost-in-slow-dashboards-15e704be95904504a750949a474fb145#1094972251824000adca4226f80d0592), while the content is loading.

npfitz commented:
Thanks @agnes-pinhanelli, I'll make those copy changes. 

> But this is still in progress, right? Asking because I wanted to check if we would be able to apply[ the animated indicator instead of just the favicon that is proposed in the product doc](https://www.notion.so/metabase/Reduce-the-waiting-cost-in-slow-dashboards-15e704be95904504a750949a474fb145#1094972251824000adca4226f80d0592), while the content is loading.

Correct! I've split this up into a few different deliverables in #21247. On the topic of the loading indicator, I've split that into it's own deliverable because I don't think it'll be trivial. Firefox will support using a .gif, but from my research it looks like I'm going to need handle it frame by frame in every other browser. Looking forward to tackling that though haha

npfitz commented:
Oh! And yes, one the user returns to the page we show the normal page title. The checkmark also returns to normal after 3 seconds of the user returning to the page

agnes-pinhanelli commented:
@npfitz [npfitz](https://github.com/npfitz) Sounds great! Thank you so much

daltojohnso reviewed (changes requested):
Nice work, a few minor change requests.

I'm wondering about the new favicon assets we are adding and if we will need to wire them up as configurable settings, since the favicon is already a configurable setting if you go to admin settings > whitelabel. I'm sure for users who've replaced the metabase logo having one appear during loading might be seen as objectionable.

npfitz committed:
favicon feedback

ranquild commented on frontend/src/metabase/dashboard/containers/DashboardApp.jsx:
```diff
@@ -93,6 +131,25 @@ export default class DashboardApp extends Component {
+        flashFavicon(LOAD_COMPLETE_FAVICON, 3000);
+      } else {
+        flashFavicon(LOAD_COMPLETE_FAVICON);
+        document.addEventListener(
```
Yep

npfitz committed:
Adding Favicon HoC, moving favicon logic to actions

ranquild commented on frontend/src/metabase/hoc/Favicon.jsx:
```diff
@@ -0,0 +1,49 @@
+import _ from "underscore";
+import MetabaseSettings from "../lib/settings";
+
+const DEFAULT_FAVICON = MetabaseSettings.get("application-favicon-url");
```
This can be changed in runtime, I believe you need to read in the the component

ranquild commented on frontend/src/metabase/hoc/Favicon.jsx:
```diff
@@ -0,0 +1,49 @@
+    .setAttribute("href", currentFavicon);
+});
+
+const favicon = faviconSetterOrGetter => ComposedComponent =>
```
Could this be written using a functional component instead? E.g.

```
const withFavicon = selector => ComposedComponent => props => {
  const favicon = selector(props);

  useEffect(() => {
    document
      .querySelector('link[rel="icon"]')
      .setAttribute("href", favicon);
  }, [favicon]);

  return <WrappedComponent  {...props}    />;
};

export default withFavicon;
```

It would be also great to use TypeScript for a new component

ranquild replied:
If there are issues with this change, could use use `componentDidMount` instead of `UNSAFE_` handler?

ranquild commented on frontend/src/metabase/dashboard/containers/DashboardApp.jsx:
```diff
@@ -74,7 +88,31 @@ const mapDispatchToProps = {
+@favicon(({ showLoadingCompleteFavicon }) =>
+  showLoadingCompleteFavicon ? LOAD_COMPLETE_FAVICON : null,
+)
+@title(
```
nit: move it to `selectors.js` or `utils.js`?

ranquild reviewed (changes requested):
Looks very good, requesting changes mostly to fix the issue of reading settings statically.

npfitz committed:
Resolving Alex's feedback

npfitz committed:
adding displayName to Favicon HoC

npfitz committed:
Removing static display name from DashbaordApp component

ranquild commented on frontend/src/metabase/hoc/Favicon.jsx:
```diff
@@ -0,0 +1,35 @@
+    const favicon = resolveFavicon(faviconSetterOrGetter, props);
+
+    useEffect(() => {
+      document.querySelector('link[rel="icon"]').setAttribute("href", favicon);
```
Do we need to a return a cleanup function here? i.e. to set the default favicon in case the component unmounts

npfitz replied:
Just added :)

ranquild reviewed (approved):
Looks good. I believe we need to add a cleanup function to the effect, but that's it. Great work!

kulyk reviewed (approved):
LGTM! Nice work! Please take a look at the last few comments from Alexander though

npfitz committed:
Adding function to useEffect hook in Favicon.jsx

npfitz commented:
I believe I've addressed all of Alexander's feedback. I just went through and resolved the conversations for comments he had. This may be good to merge!

daltojohnso reviewed (approved):
👍

npfitz commented:
retest this please
